File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        // O(n) time & O(h) space
        
        // Base case
        if(root == null) return 0;
        
        // Recursive case
        int heightOfLeftSubtree = maxDepth(root.left);
        int heightOfRightSubtree = maxDepth(root.right);
        
        return Math.max(heightOfLeftSubtree, heightOfRightSubtree) + 1;
    }
}

File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # O(n) time & O(h) space
        
        # Base case
        if root == None: return 0
        
        # Recursive case
        left_depth = self.maxDepth(root.left)
        right_depth = self.maxDepth(root.right)
        
        node_depth = max(left_depth, right_depth) + 1
        
        return node_depth